README.md
# RNOpenSSLCryptor

This project is a fork of the original RNCryptor project, because openSSL support has been removed after version 2.2.

Push new pod version protocol
=====================

1. Switch branch to openSSL
2. Update tag version in ```RNOpenSSLCryptor.podspec```
3. Update Changelog.md
3. Commit
4. Create new local tag : ```git tag RNOpenSSLCryptor-2.X -a -m "New Tag"```
5. Push tag : ```git push --tags```
6. Push master branch
7. Deploy new pod ```pod repo push youboox RNOpenSSLCryptor.podspec --use-libraries --allow-warnings```
8. Update RNOpenSSLCryptor pod in your project : ```pod update RNOpenSSLCryptor```

# RNCryptor

Cross-language AES Encryptor/Decryptor [data
format](https://github.com/rnapier/RNCryptor/wiki/Data-Format).
 
The primary target is Objective-C, but implementations are available in
[C](https://github.com/megabri/MGCryptor),
[Java](http://code.google.com/p/jncryptor/),
[PHP](https://github.com/rnapier/RNCryptor/tree/master/php),
[Python](https://github.com/rnapier/RNCryptor/blob/master/python/RNCryptor.py),
and [Ruby](https://github.com/rnapier/RNCryptor/tree/master/ruby).

The data format includes all the metadata required to securely implement AES
encryption, as described in ["Properly encrypting with AES with
CommonCrypto,"](http://robnapier.net/blog/aes-commoncrypto) and [*iOS 6
Programming Pushing the Limits*](http://iosptl.com), Chapter 15. Specifically,
it includes:

* AES-256 encryption
* CBC mode
* Password stretching with PBKDF2
* Password salting
* Random IV
* Encrypt-then-hash HMAC

## Basic Objective-C Usage

The most common in-memory use case is as follows:

``` objc
NSData *data = [@"Data" dataUsingEncoding:NSUTF8StringEncoding];
NSError *error;
NSData *encryptedData = [RNEncryptor encryptData:data
                                   	withSettings:kRNCryptorAES256Settings
                                          password:aPassword
                                             error:&error];
```

This generates an `NSData` including a header, encryption salt, HMAC salt, IV,
ciphertext, and HMAC. To decrypt this bundle:

``` objc
NSData *decryptedData = [RNDecryptor decryptData:encryptedData
                                    withPassword:aPassword
                                           error:&error];
```

Note that `RNDecryptor` does not require settings. These are read from the
header.

## Asynchronous use

`RNCryptor suports asynchronous use, specifically designed to work with
`NSURLConnection. This is also useful for cases where the encrypted or decrypted
`data will not comfortably fit in memory. If the data will comfortably fit in
`memory, asynchronous operation is best acheived using dispatch_async().

To operate in asynchronous mode, you create an `RNEncryptor` or `RNDecryptor`,
providing it a handler. This handler will be called as data is encrypted or
decrypted. As data becomes available, call `addData:`. When you reach the end of
the data call `finish`.

``` objc
- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
  [self.encryptedData addData:[self.cryptor addData:data]];
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
  [self.cryptor finish];
}

// Other connection delegates

- (void)decryptionDidFinish {
  if (self.cryptor.error) {
    // An error occurred. You cannot trust encryptedData at this point
  }
  else {
    // self.encryptedData is complete. Use it as you like
  }
  self.encryptedData = nil;
  self.cryptor = nil;
  self.connection = nil;
}

- (void)decryptRequest:(NSURLRequest *)request {
  self.encryptedData = [NSMutableData data];
  self.connection = [[NSURLConnection alloc] initWithRequest:request delegate:self];
  self.cryptor = [[RNDecryptor alloc] initWithPassword:self.password
                                               handler:^(RNCryptor *cryptor, NSData *data) {
                                                   [self.decryptedData appendData:data];
                                                   if (cryptor.isFinished) {
                                                     [self decryptionDidFinish];
                                                   }
                                                 }];
}
```

## Async and Streams

When performing async operations on streams, the data can come very quickly
(particularly if you're reading from a local file). If you use RNCryptor in a
naïve way, you'll queue a work blocks faster than the engine can process them
and your memory usage will spike. This is particularly true if there's only one
core, such as on an iPad 1. The solution is to only dispatch new work blocks as
the previous work blocks complete. When performing async operations on streams,
the data can come very quickly (particularly if you're reading from a local
file). If you use RNCryptor in a naïve way, you'll queue a work blocks faster
than the engine can process them and your memory usage will spike. This is
particularly true if there's only one core, such as on an iPad 1. The solution
is to only dispatch new work blocks as the previous work blocks complete. When
performing async operations on streams, the data can come very quickly
(particularly if you're reading from a local file). If you use RNCryptor in a
naïve way, you'll queue a work blocks faster than the engine can process them
and your memory usage will spike. This is particularly true if there's only one
core, such as on an iPad 1. The solution is to only dispatch new work blocks as
the previous work blocks complete.

``` objc
// Make sure that this number is larger than the header + 1 block.
// 33+16 bytes = 49 bytes. So it shouldn't be a problem.
int blockSize = 32 * 1024;

NSInputStream *cryptedStream = [NSInputStream inputStreamWithFileAtPath:@"C++ Spec.pdf"];
NSOutputStream *decryptedStream = [NSOutputStream outputStreamToFileAtPath:@"/tmp/C++.crypt" append:NO];

[cryptedStream open];
[decryptedStream open];

// We don't need to keep making new NSData objects. We can just use one repeatedly.
__block NSMutableData *data = [NSMutableData dataWithLength:blockSize];
__block RNEncryptor *decryptor = nil;

dispatch_block_t readStreamBlock = ^{
  [data setLength:blockSize];
  NSInteger bytesRead = [cryptedStream read:[data mutableBytes] maxLength:blockSize];
  if (bytesRead < 0) {
    // Throw an error
  }
  else if (bytesRead == 0) {
    [decryptor finish];
  }
  else {
    [data setLength:bytesRead];
    [decryptor addData:data];
    NSLog(@"Sent %ld bytes to decryptor", (unsigned long)bytesRead);
  }
};

decryptor = [[RNEncryptor alloc] initWithSettings:kRNCryptorAES256Settings
                                         password:@"blah"
                                          handler:^(RNCryptor *cryptor, NSData *data) {
                                            NSLog(@"Decryptor recevied %ld bytes", (unsigned long)data.length);
                                            [decryptedStream write:data.bytes maxLength:data.length];
                                            if (cryptor.isFinished) {
                                              [decryptedStream close];
                                              // call my delegate that I'm finished with decrypting
                                            }
                                            else {
                                              // Might want to put this in a dispatch_async(), but I don't think you need it.
                                              readStreamBlock();
                                            }
                                          }];

// Read the first block to kick things off    
readStreamBlock();
```

I'll eventually get this into the API to simplify things. See [Cyrille's SO
question](http://stackoverflow.com/a/14586231/97337) for more discussion. Pull
requests welcome.

## Building

Comes packaged as a static library, but the source files can be dropped into any
project. The OpenSSL files are not required.

Requires `Security.framework`.

Supports 10.6+ and iOS 4+.

The current file format is v3. To read v1 files (see Issue #44), you need to set the compile-time macro `RNCRYPTOR_ALLOW_V1_BAD_HMAC`. It is not possible to write v1 files anymore.

## Design considerations

`RNCryptor` has several design goals, in order of importance:

### Easy to use correctly for most common use cases

The most critical concern is that it be easy for non-experts to use `RNCryptor` correctly. A framework that is more secure, but requires a steep learning curve on the developer will either be not used, or used incorrectly. Whenever possible, a single line of code should "do the right thing" for the most common cases.

This also requires that it fail correctly and provide good errors.

### Reliance on CommonCryptor functionality

`RNCryptor` has very little "security" code. It relies as much as possible on the OS-provided CommonCryptor. If a feature does not exist in CommonCryptor, then it generally will not be provided in `RNCryptor`.

### Best practice security

Wherever possible within the above constraints, the best available algorithms
are applied. This means AES-256, HMAC+SHA1, and PBKDF2:

* AES-256. While Bruce Schneier has made some interesting recommendations
regarding moving to AES-128 due to certain attacks on AES-256, my current
thinking is in line with [Colin
Percival](http://www.daemonology.net/blog/2009-07-31-thoughts-on-AES.html).
PBKDF2 output is effectively random, which should negate related-keys attacks
against the kinds of use cases we're interested in.

* AES-CBC mode. This was a somewhat complex decision, but the ubiquity of CBC
outweighs other considerations here. There are no major problems with CBC mode,
and nonce-based modes like CTR have other trade-offs. See ["Mode changes for
RNCryptor"](http://robnapier.net/blog/mode-rncryptor) for more details on this
decision.

* Encrypt-then-MAC. If there were a good authenticated AES mode on iOS (GCM for
instance), I would probably use that for its simplicity. Colin Percival makes
[good arguments for hand-coding an encrypt-than-
MAC](http://www.daemonology.net/blog/2009-06-24-encrypt-then-mac.html) rather
than using an authenticated AES mode, but in RNCryptor mananging the HMAC
actually adds quite a bit of complexity. I'd rather the complexity at a more
broadly peer-reviewed layer like CommonCryptor than at the RNCryptor layer. But
this isn't an option, so I fall back to my own Encrypt-than-MAC.

* HMAC+SHA256. No surprises here.

* PBKDF2. While bcrypt and scrypt may be more secure than PBKDF2, CommonCryptor
only supports PBKDF2. [NIST also continues to recommend
PBKDF2](http://security.stackexchange.com/questions/4781/do-any-security-
experts-recommend-bcrypt-for-password-storage). We use 10k rounds of PBKDF2
which represents about 80ms on an iPhone 4.

### Code simplicity

`RNCryptor endeavors to be implemented as simply as possible, avoiding tricky
`code. It is designed to be easy to read and code review.

### Performance

Performance is a goal, but not the most important goal. The code must be secure
and easy to use. Within that, it is as fast and memory-efficient as possible.

### Portability

Without sacrificing other goals, it is preferable to read the output format of
`RNCryptor` on other platforms.

## Version History

* v2.2 Switches to file format v3 to deal with Issue #77.
* v2.1 Switches to file format v2 to deal with Issue #44.
* v2.0 adds asynchronous modes.
* v2.1 backports `RNCryptor` to older versions of Mac OS X (and possibly iOS).


## LICENSE

Except where otherwise indicated in the source code, this code is licensed under
the MIT License:

```
Permission is hereby granted, free of charge, to any person obtaining a 
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:
 
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
```

Portions of this code, indicated in the source, are licensed under the following
license:

```
/*-
 * Copyright (c) 2008 Damien Bergamini <damien.bergamini@free.fr>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
```

Portions of ths code, indicated in the source, are licensed under the APSL
license:

```
/*
 * Copyright (c) 2006-2010 Apple Inc. All Rights Reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 *
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 *
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 *
 * @APPLE_LICENSE_HEADER_END@
 */
```


Changelog.md
# Version 2.3

Remove unnecessary use of the C function `SecRandomCopyBytes` because it produce a Duplicate Symbol error since Xcode 9.

# Version 2.2

Version 2.2 is a fairly large release. It's been almost a year since 2.1 came out, and there are many small and large bug fixes.

V2.2 updates the file format from 2 to 3. It will read format 2 files, but will only write format 3. These are not readable by RNCryptor v2.1. See Issue #77 for details. The PHP, Python, and Ruby implementations also write format 3 and read format 2 or 3.

## Security Issues

* Issue #91:  Use constant time HMAC comparisons to avoid timing attacks
* Issue #77: KeyForPassword() broken for multi-byte passwords (UTF-8)

## Other Noteable Changes

* Improved PHP, Python, and Ruby implementations
* Improved test cases, with test vectors
* Issue #76: Support OSX in podspec
* Resolved static analyzer warnings
* Ensure compatibility with iOS 4.2
* Accept settings to RNDecryptor (Issue #65)
* Copy password rather than retain it (Issue #64)
* Crash when reading v1 header


RNCryptor/RNCryptor.m
//
//  RNCryptor.m
//
//  Copyright (c) 2012 Rob Napier
//
//  This code is licensed under the MIT License:
//
//  Permission is hereby granted, free of charge, to any person obtaining a
//  copy of this software and associated documentation files (the "Software"),
//  to deal in the Software without restriction, including without limitation
//  the rights to use, copy, modify, merge, publish, distribute, sublicense,
//  and/or sell copies of the Software, and to permit persons to whom the
//  Software is furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
//  DEALINGS IN THE SOFTWARE.
//
//
#import "RNCryptor.h"
#import "RNCryptor+Private.h"
#import <Security/SecRandom.h>

extern int
CCKeyDerivationPBKDF( CCPBKDFAlgorithm algorithm, const char *password, size_t passwordLen,
                     const uint8_t *salt, size_t saltLen,
                     CCPseudoRandomAlgorithm prf, uint rounds,
                     uint8_t *derivedKey, size_t derivedKeyLen) __attribute__((weak_import));


NSString *const kRNCryptorErrorDomain = @"net.robnapier.RNCryptManager";
const uint8_t kRNCryptorFileVersion = 3;

// TODO: This is a slightly expensive solution, but it's convenient. May want to create a "walkable" data object
@implementation NSMutableData (RNCryptor)
- (NSData *)_RNConsumeToIndex:(NSUInteger)index
{
    NSData *removed = [self subdataWithRange:NSMakeRange(0, index)];
    [self replaceBytesInRange:NSMakeRange(0, self.length - index) withBytes:([self mutableBytes] + index)];
    [self setLength:self.length - index];
    return removed;
}
@end


@implementation RNCryptor
@synthesize responseQueue = _responseQueue;
@synthesize engine = _engine;
@synthesize outData = __outData;
@synthesize queue = _queue;
@synthesize HMACLength = __HMACLength;
@synthesize error = _error;
@synthesize finished = _finished;
@synthesize options = _options;
@synthesize handler = _handler;

+ (NSData *)synchronousResultForCryptor:(RNCryptor *)cryptor data:(NSData *)inData error:(NSError **)anError
{
    dispatch_semaphore_t sem = dispatch_semaphore_create(0);
    
    NSMutableData *data = [NSMutableData data];
    __block NSError *returnedError = nil;
    
    RNCryptorHandler handler = ^(RNCryptor *c, NSData *d) {
        [data appendData:d];
        if (c.isFinished) {
            returnedError = c.error;
            dispatch_semaphore_signal(sem);
        }
    };
    
    cryptor.handler = handler;
    
    dispatch_queue_t queue = dispatch_queue_create("net.robnapier.RNEncryptor.response", DISPATCH_QUEUE_SERIAL);
    cryptor.responseQueue = queue;
    [cryptor addData:inData];
    [cryptor finish];
    
    
    dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
    
#if !OS_OBJECT_USE_OBJC
    dispatch_release(sem);
    if (queue) {
        dispatch_release(queue);
    }
#endif
    
    if (returnedError) {
        if (anError) {
            *anError = returnedError;
        }
        return nil;
    }
    else {
        return data;
    }
}

// For use with OS X 10.6
// Based on http://opensource.apple.com/source/CommonCrypto/CommonCrypto-55010/Source/API/CommonKeyDerivation.c
/*-
 * Copyright (c) 2008 Damien Bergamini <damien.bergamini@free.fr>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#define CC_MAX_PRF_WORKSPACE 128+4
#define kCCPRFHmacAlgSHA1hlen    CC_SHA1_DIGEST_LENGTH
#define kCCPRFHmacAlgSHA224hlen CC_SHA224_DIGEST_LENGTH
#define kCCPRFHmacAlgSHA256hlen CC_SHA256_DIGEST_LENGTH
#define kCCPRFHmacAlgSHA384hlen CC_SHA384_DIGEST_LENGTH
#define kCCPRFHmacAlgSHA512hlen CC_SHA512_DIGEST_LENGTH

static size_t
getPRFhlen(CCPseudoRandomAlgorithm prf)
{
    switch(prf) {
        case kCCPRFHmacAlgSHA1:        return kCCPRFHmacAlgSHA1hlen;
        case kCCPRFHmacAlgSHA224:    return kCCPRFHmacAlgSHA224hlen;
        case kCCPRFHmacAlgSHA256:    return kCCPRFHmacAlgSHA256hlen;
        case kCCPRFHmacAlgSHA384:    return kCCPRFHmacAlgSHA384hlen;
        case kCCPRFHmacAlgSHA512:    return kCCPRFHmacAlgSHA512hlen;
        default:
            NSCAssert(NO, @"Unknown prf: %d", prf);
            return 1;
    }
}

static void
PRF(CCPseudoRandomAlgorithm prf, const char *password, size_t passwordLen, u_int8_t *salt, size_t saltLen, u_int8_t *output)
{
    switch(prf) {
        case kCCPRFHmacAlgSHA1:
            CCHmac(kCCHmacAlgSHA1, password, passwordLen, salt, saltLen, output);
            break;
        case kCCPRFHmacAlgSHA224:
            CCHmac(kCCHmacAlgSHA224, password, passwordLen, salt, saltLen, output);
            break;
        case kCCPRFHmacAlgSHA256:
            CCHmac(kCCHmacAlgSHA256, password, passwordLen, salt, saltLen, output);
            break;
        case kCCPRFHmacAlgSHA384:
            CCHmac(kCCHmacAlgSHA384, password, passwordLen, salt, saltLen, output);
            break;
        case kCCPRFHmacAlgSHA512:
            CCHmac(kCCHmacAlgSHA512, password, passwordLen, salt, saltLen, output);
            break;
    }
}

static int
RN_CCKeyDerivationPBKDF( CCPBKDFAlgorithm algorithm, const char *password, size_t passwordLen,
                        const uint8_t *salt, size_t saltLen,
                        CCPseudoRandomAlgorithm prf, uint rounds,
                        uint8_t *derivedKey, size_t derivedKeyLen)
{
    u_int8_t oldbuffer[CC_MAX_PRF_WORKSPACE], newbuffer[CC_MAX_PRF_WORKSPACE],
    saltCopy[CC_MAX_PRF_WORKSPACE+4], collector[CC_MAX_PRF_WORKSPACE];
    int rawblock, i, j;
    size_t r, nblocks;
    size_t    hlen, offset;
    
    if(algorithm != kCCPBKDF2) return -1;
    
    /*
     * Check initial parameters
     */
    
    if (rounds < 1 || derivedKeyLen == 0)
        return -1; // bad parameters
    if (saltLen == 0 || saltLen > CC_MAX_PRF_WORKSPACE)
        return -1; // out of bounds parameters
    
    hlen = getPRFhlen(prf);
    
    /*
     * FromSpec: Let l be the number of hLen-octet blocks in the derived key, rounding up,
     * and let r be the number of octets in the last block:
     */
    
    nblocks = (derivedKeyLen+hlen-1)/hlen; // in the spec nblocks is referred to as l
    r = derivedKeyLen % hlen;
    r = (r) ? r: hlen;
    
    /*
     * Make a copy of the salt buffer so we can concatenate the
     * block counter for each series of rounds.
     */
    
    memcpy(saltCopy, salt, saltLen);
    bzero(derivedKey, derivedKeyLen);
    
    /*
     * FromSpec:
     *
     * For each block of the derived key apply the function F defined below to the password P,
     * the salt S, the iteration count c, and the block index to compute the block:
     *
     *           F(P,S,c,i)=U1 \xorU2 \xor⋅⋅⋅\xorUc
     *
     * where
     *                U1 =PRF(P,S||INT (i)),
     *                U2 =PRF(P,U1),
     *                ...
     *                Uc = PRF (P, Uc-1) .
     */
    
    for(rawblock = 0; rawblock < nblocks; rawblock++) {
        int block = rawblock+1;
        size_t copyLen;
        
        offset = rawblock * hlen;
        copyLen = (block != nblocks) ? hlen: r;
        
        /*
         * FromSpec: Here, INT (i) is a four-octet encoding of the integer i, most significant octet first.
         */
        
        for(i=0; i<4; i++) saltCopy[saltLen+i] = (block >> 8*(3-i)) & 0xff;
        
        PRF(prf, password, passwordLen, saltCopy, saltLen+4, oldbuffer);                    // Initial PRF with the modified salt
        
        memcpy(collector, oldbuffer, hlen);                                                // Initial value for this block of the derived key.
        
        for(i = 1; i < rounds; i++) {
            PRF(prf, password, passwordLen, oldbuffer, hlen, newbuffer);                        // Subsequent PRF with the previous result as the salt
            memcpy(oldbuffer, newbuffer, hlen);
            for(j = 0; j < hlen; j++) collector[j] ^= newbuffer[j];                    // Xoring the round into the collector
        }
        memcpy(derivedKey+offset, collector, copyLen);
    }
    
    /*
     * Clear temp buffers.
     */
    
    bzero(oldbuffer, CC_MAX_PRF_WORKSPACE);
    bzero(newbuffer, CC_MAX_PRF_WORKSPACE);
    bzero(collector, CC_MAX_PRF_WORKSPACE);
    bzero(saltCopy, CC_MAX_PRF_WORKSPACE+4);
    
    return 0;
}

/* End code derived from CommonKeyDerivation.c */


+ (NSData *)keyForPassword:(NSString *)password salt:(NSData *)salt settings:(RNCryptorKeyDerivationSettings)keySettings
{
    NSMutableData *derivedKey = [NSMutableData dataWithLength:keySettings.keySize];
    
    // See Issue #77. V2 incorrectly calculated key for multi-byte characters.
    NSData *passwordData;
    if (keySettings.hasV2Password) {
        passwordData = [NSData dataWithBytes:[password UTF8String] length:[password length]];
    }
    else {
        passwordData = [password dataUsingEncoding:NSUTF8StringEncoding];
    }
    
    // Use the built-in PBKDF2 if it's available. Otherwise, we have our own. Hello crazy function pointer.
    int result;
    int (*PBKDF)(CCPBKDFAlgorithm algorithm, const char *password, size_t passwordLen,
                 const uint8_t *salt, size_t saltLen,
                 CCPseudoRandomAlgorithm prf, uint rounds,
                 uint8_t *derivedKey, size_t derivedKeyLen);
    
    PBKDF = CCKeyDerivationPBKDF ?: RN_CCKeyDerivationPBKDF;
    
    result = PBKDF(keySettings.PBKDFAlgorithm,         // algorithm
                   passwordData.bytes,                 // password
                   passwordData.length,                // passwordLength
                   salt.bytes,                         // salt
                   salt.length,                        // saltLen
                   keySettings.PRF,                    // PRF
                   keySettings.rounds,                 // rounds
                   derivedKey.mutableBytes,            // derivedKey
                   derivedKey.length);                 // derivedKeyLen
    
    // Do not log password here
    NSAssert(result == kCCSuccess, @"Unable to create AES key for password: %d", result);
    
    return derivedKey;
}

// For use on OS X 10.6
// Based on http://www.opensource.apple.com/source/Security/Security-55179.1/sec/Security/SecFramework.c
// Modified by Rob Napier April, 2013.
/*
 * Copyright (c) 2006-2010 Apple Inc. All Rights Reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 *
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 *
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 *
 * @APPLE_LICENSE_HEADER_END@
 */
static int RN_SecRandomCopyBytes(void *rnd, size_t count, uint8_t *bytes) {
    static int kSecRandomFD;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        kSecRandomFD = open("/dev/random", O_RDONLY);
    });
    
    if (kSecRandomFD < 0)
        return -1;
    while (count) {
        ssize_t bytes_read = read(kSecRandomFD, bytes, count);
        if (bytes_read == -1) {
            if (errno == EINTR)
                continue;
            return -1;
        }
        if (bytes_read == 0) {
            return -1;
        }
        bytes += bytes_read;
        count -= bytes_read;
    }
    
    return 0;
}
/* End code dervied from SecFramework.c */

+ (NSData *)randomDataOfLength:(size_t)length
{
    NSMutableData *data = [NSMutableData dataWithLength:length];
    
    int result = RN_SecRandomCopyBytes(NULL, length, data.mutableBytes);
    NSAssert(result == 0, @"Unable to generate random bytes: %d", errno);
    
    return data;
}

- (id)initWithHandler:(RNCryptorHandler)handler
{
    NSParameterAssert(handler);
    self = [super init];
    if (self) {
        NSString *responseQueueName = [@"net.robnapier.response." stringByAppendingString:NSStringFromClass([self class])];
        _responseQueue = dispatch_queue_create([responseQueueName UTF8String], NULL);
        
        NSString *queueName = [@"net.robnapier." stringByAppendingString:NSStringFromClass([self class])];
        _queue = dispatch_queue_create([queueName UTF8String], DISPATCH_QUEUE_SERIAL);
        __outData = [NSMutableData data];
        
        _handler = [handler copy];
    }
    return self;
}

- (void)dealloc
{
    if (_responseQueue) {
#if !OS_OBJECT_USE_OBJC
        dispatch_release(_responseQueue);
#endif
        _responseQueue = NULL;
    }
    
    if (_queue) {
#if !OS_OBJECT_USE_OBJC
        dispatch_release(_queue);
#endif
        _queue = NULL;
    }
}

- (void)setResponseQueue:(dispatch_queue_t)aResponseQueue
{
    if (aResponseQueue) {
#if !OS_OBJECT_USE_OBJC
        dispatch_retain(aResponseQueue);
#endif
    }
    
    if (_responseQueue) {
#if !OS_OBJECT_USE_OBJC
        dispatch_release(_responseQueue);
#endif
    }
    
    _responseQueue = aResponseQueue;
}

- (void)addData:(NSData *)data
{
    
}

- (void)finish
{
    
}

- (void)cleanupAndNotifyWithError:(NSError *)error
{
    self.error = error;
    self.finished = YES;
    if (self.handler) {
        dispatch_sync(self.responseQueue, ^{
            self.handler(self, self.outData);
        });
        self.handler = nil;
    }
}

- (BOOL)hasHMAC
{
    return self.HMACLength > 0;
}


@end

